### PR TITLE
Support union field

### LINKED ISSUE #63: Adding support for dataclass Union field
Hi there!

We are starting using this package a lot in our product and it's great!
One gap we reached into is support for dataclass union (`typing.Union[dataclass1, dataclass2, ...]`) as one of the dataclass fields.

Example case:

```python
@dataclass
class A:
    a: int


@dataclass
class B:
    b: int


@dataclass
class Response:
    obj: A | B

class ResponseSerializer(DataclassSerializer)
    class Meta:
        dataclass = Response
```

I solved this by extending `DataclassSerializer` and add support using `rest-polymorphic`.
We also use `drf-spectacular` for `openapi` scheme and it also supports `rest-polymorphic`, so all good!

Before I open PR I wanted to check that this feature is useful, I would be happy to get a feedback :)

### PR DESCRIPTION
**Pull request by panchock**
Resolves #63

### CODE REVIEW

File: rest_framework_dataclasses/serializers.py [hunk 1]
```diff
@@ -149,7 +150,7 @@ def dataclass_definition(self) -> DataclassDefinition[T]:
             dataclass_type = meta.dataclass
 
         # Make sure we're dealing with an actual dataclass.
-        if not dataclasses.is_dataclass(dataclass_type):
+        if not dataclasses.is_dataclass(cast(Type[T], dataclass_type)):
```
Comment (panchock) [inline rest_framework_dataclasses/serializers.py:153]: I had mypy error here and wasn't sure why so meanwhile I casted it.
https://github.com/oxan/djangorestframework-dataclasses/actions/runs/4707423442/jobs/8349271383

Comment (oxan) [reply rest_framework_dataclasses/serializers.py:153]: I think this is a result of mypy updating and changing their rules, and unrelated to this PR... I'll look into it.

File: rest_framework_dataclasses/union_serializer.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    This class utlizes PolymorphicSerializer from rest_polymorphic packageto support
+    any class object (not limited to django models).
+
+    Creating PolymorphicSerializer requires declaring a new class derived from it with "model_serializer_mapping"
+    classvar.
+    We could create dynamically new class derived from PolymorphicSerializer with "model_serializer_mapping" classvar,
+    but in case that we have 2 classes with the same name (for example, same union in two different serializer)
+    drf-spectacular warns that there is a conflict (2 different classes with the same name).
+    In order to avoid this warning we create new class that let us get the "model_serializer_mapping"
+    classvar (and few others classvars) from the constructor and actually make them instance members.
```
Comment (oxan) [inline rest_framework_dataclasses/union_serializer.py:31]: I'm not sure I fully understand this reasoning: I get that `drf-spectacular` doesn't like it if there's two serializer classes with the same name, but I don't see how creating this custom class prevents that.

**Review by oxan — COMMENTED** [review]
I like it! I've one question though, see below.

**Comment by oxan** [comment]
I've decided to implement union type support using a homegrown `UnionField` (#82) instead of a solution based on this PR/`django-rest-polymorphic`. The main reason for that is that such a field turns out to be less code than the `InlinePolymorphicSerializer` needed here to fixup the mismatch with `django-rest-polymorphic` assumptions, while also being more flexible (it supports e.g. unions of primitives). It also avoids the dependency.

**Comment by panchock** [comment]
> I've decided to implement union type support using a homegrown `UnionField` (#82) instead of a solution based on this PR/`django-rest-polymorphic`. The main reason for that is that such a field turns out to be less code than the `InlinePolymorphicSerializer` needed here to fixup the mismatch with `django-rest-polymorphic` assumptions, while also being more flexible (it supports e.g. unions of primitives). It also avoids the dependency.
> 
> 

@oxan This solution also support drf spectacular? What will happens if someone will try to generate openapi schema now?

**Comment by oxan** [comment]
I don't know, I haven't tested it as I don't use drf-spectacular. However, if it doesn't work, the best way to fix it would be by extending drf-spectactular's [rest_framework_dataclasses](https://github.com/tfranzel/drf-spectacular/blob/master/drf_spectacular/contrib/rest_framework_dataclasses.py) plugin.

**Pull request closed without merging**